frame/elections-phragmen/src/lib.rs

```diff
@@ -406,55 +446,51 @@ decl_module! {
 		///   Similar to [`remove_voter`], if replacement runners exists, they are immediately used.
 		#[weight = (2_000_000_000, DispatchClass::Operational)]
-		fn renounce_candidacy(origin) {
+		fn renounce_candidacy(origin, renouncing: Renouncing) {
 			let who = ensure_signed(origin)?;
-
-			// NOTE: this function attempts the 3 conditions (being a candidate, member, runner) and
-			// fails if none are matched. Unlike other Palette functions and modules where checks
-			// happen first and then execution happens, this function is written the other way
-			// around. The main intention is that reading all of the candidates, members and runners
-			// from storage is expensive. Furthermore, we know (soft proof) that they are always
-			// mutually exclusive. Hence, we try one, and only then decode more storage.
-
-			if let Ok(_replacement) = Self::remove_and_replace_member(&who) {
-				T::Currency::unreserve(&who, T::CandidacyBond::get());
-				Self::deposit_event(RawEvent::MemberRenounced(who.clone()));
-
-				// safety guard to make sure we do only one arm. Better to read runners later.
-				return Ok(());
-			}
-
-			let mut runners_up_with_stake = Self::runners_up();
-			if let Some(index) = runners_up_with_stake.iter()
-				.position(|(ref r, ref _s)| r == &who)
-			{
-				runners_up_with_stake.remove(index);
-				// unreserve the bond
-				T::Currency::unreserve(&who, T::CandidacyBond::get());
-				// update storage.
-				<RunnersUp<T>>::put(runners_up_with_stake);
-				// safety guard to make sure we do only one arm. Better to read runners later.
-				return Ok(());
-			}
-
-			let mut candidates = Self::candidates();
-			if let Ok(index) = candidates.binary_search(&who) {
-				candidates.remove(index);
-				// unreserve the bond
-				T::Currency::unreserve(&who, T::CandidacyBond::get());
-				// update storage.
-				<Candidates<T>>::put(candidates);
-				// safety guard to make sure we do only one arm. Better to read runners later.
-				return Ok(());
-			}
-
-			Err(Error::<T>::InvalidOrigin)?
+			match renouncing {
+				Renouncing::Member => {
+					// returns NorMember error in case of error.
+					let _ = Self::remove_and_replace_member(&who)?;
+					T::Currency::unreserve(&who, T::CandidacyBond::get());
+					Self::deposit_event(RawEvent::MemberRenounced(who.clone()));
+				},
+				Renouncing::RunnerUp => {
+					let mut runners_up_with_stake = Self::runners_up();
+					if let Some(index) = runners_up_with_stake
+						.iter()
+						.position(|(ref r, ref _s)| r == &who)
+					{
+						runners_up_with_stake.remove(index);
+						// unreserve the bond
+						T::Currency::unreserve(&who, T::CandidacyBond::get());
+						// update storage.
+						<RunnersUp<T>>::put(runners_up_with_stake);
+					} else {
+						Err(Error::<T>::InvalidRenouncing)?;
+					}
+				}
+				Renouncing::Candidate(index) => {
+					let mut candidates = Self::candidates();
```

Comment:
are candidates (and runners up for that matter) bounded? otherwise you might want the caller to pass in counts to estimate weight

Reply:
runners-up are but candidates are not. Yeah this one could use the length of candidates.

Reply:
I'll probably end up doing this anyway in the main branch. Thanks for noting.